sourcerer-core/src/main/java/org/elder/sourcerer2/ExpectedVersion.kt

```diff
@@ -0,0 +1,127 @@
+package org.elder.sourcerer2
+
+import org.elder.sourcerer2.exceptions.ConflictingExpectedVersionsException
+
+/**
+ * Describes the version that an aggregate is expected to be in when read, or appended to.
+ */
+sealed class ExpectedVersion {
+    data class Exactly(val streamVersion: StreamVersion) : ExpectedVersion()
+    object NotCreated : ExpectedVersion() {
+        override fun toString(): String {
+            return "ExpectedVersion.NotCreated"
+        }
+    }
+    object AnyExisting : ExpectedVersion()  {
+        override fun toString(): String {
+            return "ExpectedVersion.AnyExisting"
+        }
+    }
+    object Any : ExpectedVersion()  {
+        override fun toString(): String {
+            return "ExpectedVersion.Any"
+        }
+    }
+
+    companion object {
+        // Java convenience methods.
+        @JvmStatic
+        fun any(): ExpectedVersion {
+            return ExpectedVersion.Any
+        }
+
+        @JvmStatic
+        fun anyExisting(): ExpectedVersion {
+            return ExpectedVersion.AnyExisting
+        }
+
+        @JvmStatic
+        fun notCreated(): ExpectedVersion {
+            return ExpectedVersion.NotCreated
+        }
+
+        @JvmStatic
+        fun exactly(version: StreamVersion): ExpectedVersion {
+            return ExpectedVersion.Exactly(version)
+        }
+
+        @JvmStatic
+        fun merge(
+                baseVersion: ExpectedVersion?,
+                newVersion: ExpectedVersion?
+        ): ExpectedVersion {
+            return if (baseVersion != null && newVersion != null) {
+                if (baseVersion.ordinal() > newVersion.ordinal()) {
+                    mergeOrdered(newVersion, baseVersion)
+                } else {
+                    mergeOrdered(baseVersion, newVersion)
+                }
+            } else baseVersion ?: (newVersion ?: ExpectedVersion.any())
+        }
+
+        /**
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Comments for private functions should be avoided. Prefer giving the function an expressive name. Split it up in smaller, self-explaining functions if necessary. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3ACommentOverPrivateFunction)